### PR TITLE
Implement GitHub Enterprise authn provider

### LINKED ISSUE #1793: Its possible connect to github enterprise
Hi, its possible configure extension with different github. (Use of github enterprise)

### LINKED ISSUE #112043: GitHub Enterprise support for the GitHub auth provider
Issue Type: <b>Feature Request</b>

Currently we don't support signing into GitHub Enterprise accounts using the auth provider. The auth provider makes calls to get user information that point specifically to github.com.

Either the `getSessions` API needs a way to provide the enterprise server endpoint, or the auth provider extension should automatically do this detection and let the user choose when performing sign in. The auth provider needs to store what endpoint tokens are associated with and use this instead of always going to github.com. It should also indicate this in the UI.

VS Code version: Code 1.51.1 (e5a624b788d92b8d34d1392e4c4d9789406efe8f, 2020-11-11T01:11:34.018Z)
OS version: Darwin x64 18.7.0

### PR DESCRIPTION
**Pull request by kabel**
👋 _Note from the maintainers:_ If you arrived here looking for _GitHub Enterprise Server_ support in the GitHub Pull Requests and Issues extension, please see [the wiki](https://github.com/microsoft/vscode-pull-request-github/wiki#2-is-github-enterprise-supported). **GitHub Enterprise is now tested and officially supported** (since 0.51).

<hr />

Adds an additional provider to the github-authentication extension that leverages the same code with minor modifications. A configuration item has been added for users to provide the URI of their GHE instance.

This allows GitHub API related extensions to easily integrate with
either GitHub.com or their on-premise installations.

This PR fixes microsoft/vscode-pull-request-github#1793
Also fixes #112043

Testing requires access to a GitHub Enterprise instance.

### CODE REVIEW

**Comment by kabel** [comment]
Config added for instance URI.

image

**Comment by kabel** [comment]
Successful local test.

image

**Comment by mew1033** [comment]
Awesome!

File: extensions/github-authentication/src/extension.ts
Comment (kabel) [inline extensions/github-authentication/src/extension.ts:18]: Probably need to `.bind(loginService)` here.

File: extensions/github-authentication/src/extension.ts
Comment (kabel) [inline extensions/github-authentication/src/extension.ts:23]: Probably need to `.bind(enterpriseLoginService)` here.

**Comment by lupalby** [comment]
I was able to test the changes using the configuration [here](https://github.com/microsoft/vscode-pull-request-github/issues/1793#issuecomment-778269272) and I can confirm that it works very well with the GHE instance we use at work.
Great work!

**Comment by jasonwilliams** [comment]
I can also confirm that after following the instructions from https://github.com/microsoft/vscode-pull-request-github/issues/1793#issuecomment-778269272 this with our GitHub Enterprise Server.

**Comment by RMacfarlane** [comment]
Thanks @kabel for picking this up, this is really great! I haven't had time to review this yet, but getting this in is on my to do list

**Comment by vortechs2000** [comment]
I can also confirm that after following the instructions from microsoft/vscode-pull-request-github#1793 (comment) this works with our GitHub Enterprise Server.
* GHE: 2.22
* VSCode 1.53.2
* macOS 10.15.7

File: extensions/github-authentication/src/githubServer.ts
Comment (kabel) [inline extensions/github-authentication/src/githubServer.ts:282]: Assume this would be valuable information to see where users are coming from. I don't know the schema for GDPR, so adjust as necessary.

**Comment by vortechs2000** [comment]
@RMacfarlane what needs to happen to get this and #2468 pulled in?

**Comment by RMacfarlane** [comment]
I just need to review and merge them, I am holding off on doing this since we are in the middle of our stabilization period for the next release where we do not take in new features. We should create our release branch on Friday and I will review after that

**Comment by vortechs2000** [comment]
Awesome - very excited for this to get it.  Feel free to ping me if you need something validated against a GHE 2.22 server and I'll do my best to help.

**Comment by eschweighofer** [comment]
I was able to test the changes using https://github.com/microsoft/vscode-pull-request-github/issues/1793 and confirm it works very well with our GHE instance 
Great work!
@RMacfarlane could this be merged?

**Comment by kubukoz** [comment]
Bumping, I would also love if this got merged and soon released 🙏

**Comment by jasonwilliams** [comment]
@RMacfarlane is there space in this iteration to review and merge this?

**Comment by Ygilany** [comment]
@RMacfarlane any plans to merge this soon?

**Comment by mew1033** [comment]
I'm super excited for this feature. I use GHE every day for work, and will use the pull request extension so much more when I can pair it with GHE. Can't wait to see this merged!

**Review by RMacfarlane — APPROVED** [review]
Thank you @kabel!

**Comment by catskul** [comment]
As of the new PR merge what's the right way to configure this?

**Comment by AlexDias95** [comment]
I have the same question, how can I configure the GitHub enterprise?

**Comment by kabel** [comment]
> As of the new PR merge what's the right way to configure this?

> I have the same question, how can I configure the GitHub enterprise?

I would imagine there will be some updated wiki, documentation once this feature bakes into the official release process. That said, it's only available through the _[Insiders](https://code.visualstudio.com/insiders)_ version 1.57 from 2021-05-10 or later. 

If your question resolves around how this integrates with the _[GitHub Pull Requests and Issues](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github)_ extension, you will also need to use the _[nightly build](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github-insiders)_ until an official release. Assuming you have the right version of things installed, you can set the configuration as shown in [step 4 of my preview releases](https://github.com/microsoft/vscode-pull-request-github/issues/1793#issuecomment-778269272).

In all cases, the authentication to Enterprise relies on a Personal Access Token (PAT), that must be manually generated and pasted into the appropriate prompt when you interact with a configured enterprise remote repository. Example: `https://domain-of-ghe.your-enterprise.tld/settings/tokens`

**Comment by AlexDias95** [comment]
@kabel Thanks for the help! I have installed the nightly build version and I have put in settings.json ("github-enterprise.uri": "https://github.XYZ.com",) but how can I add my token to this configuration?

**Comment by kabel** [comment]
The token is stored in a secure credential manager. It cannot be loaded through configuration; only the UI:

1. Open a code instance with a workspace/folder that contains a GHE cloned repository. Default remote name `origin` works out-of-box.
2. If you have the URI set appropriately (matches clone URI authority), the UI should prompt you that you have an account to authenticate to: image
3. Click the Accounts icon and select the "Sign in..." option: image
4. The UI will prompt for the PAT and display a status bar item:
image
image
5. Assuming PAT works well and complete authn, the UI updates the Accounts menu:
image
6. If the GitHub UI view is still _stale_ (didn't trigger the proper events), reload the window:
image
image
7. Success!
image

**Comment by jasonwilliams** [comment]
> The token is stored in a secure credential manager. It cannot be loaded through configuration; only the UI:
> 
> 1. Open a code instance with a workspace/folder that contains a GHE cloned repository. Default remote name `origin` works out-of-box.
> 2. If you have the URI set appropriately (matches clone URI authority), the UI should prompt you that you have an account to authenticate to: image
> 3. Click the Accounts icon and select the "Sign in..." option: image
> 4. The UI will prompt for the PAT and display a status bar item:
>    image
>    image
> 5. Assuming PAT works well and complete authn, the UI updates the Accounts menu:
>    image
> 6. If the GitHub UI view is still _stale_ (didn't trigger the proper events), reload the window:
>    image
>    image
> 7. Success!
>    image

@RMacfarlane You could copy this straight into the wiki 😄

**Comment by vortechs2000** [comment]
I grabbed the latest Insider build, and set `"github-enterprise.uri": "https://git.my.ghe.uri"`, but I'm not able to log into my GHE instance - I only get the option to log into github.com.  Is there some other setting I need to adjust?

**Comment by catskul** [comment]
Make sure you have the nightly of the extension as well `vscode-pull-request-github-insiders`
Also I had to restart vscode a few times before I got it to work. Also nightly stores it's config in a separate location so if you added it to vscode, and then tried it on nightly, you might be missing the config there.

**Comment by vortechs2000** [comment]
> Make sure you have the nightly of the extension as well `vscode-pull-request-github-insiders`

That did it - thank you @catskul!  I knew I was missing something obvious.

**Comment by jasonwilliams** [comment]
@connor4312 @rzhao271 

* Installed VSCode Insiders (Version: 1.57.0-insider (Universal) Commit: 387e2f39ebbbf50d4b488f900b2a745972764015)
* Installed pull request github nightly (Version: 2021.6.1983)
* I set my enterprise URI, filling out "github-enterprise.uri" (I included the schema)
* Then needed to restart VSCode
* Then i signed in following @kabel's instructions here:
https://github.com/microsoft/vscode/pull/115940#issuecomment-838838680

I can see the list of PRs but clicking on the dropdown shows nothing. 

Extension output is as follows:
```
[Info] Failed to get pull request review comments: Error: GraphQL error: Field 'path' doesn't exist on type 'PullRequestReviewThread'
GraphQL error: Field 'isOutdated' doesn't exist on type 'PullRequestReviewThread'
[Info] Error: GraphQL error: Field 'viewerViewedState' doesn't exist on type 'PullRequestChangedFile'
[Info] Failed to get pull request review comments: Error: GraphQL error: Field 'path' doesn't exist on type 'PullRequestReviewThread'
GraphQL error: Field 'isOutdated' doesn't exist on type 'PullRequestReviewThread'
[Info] Error: GraphQL error: Field 'viewerViewedState' doesn't exist on type 'PullRequestChangedFile'
[Info] Failed to get pull request review comments: Error: GraphQL error: Field 'path' doesn't exist on type 'PullRequestReviewThread'
GraphQL error: Field 'isOutdated' doesn't exist on type 'PullRequestReviewThread'
[Info] Error: GraphQL error: Field 'viewerViewedState' doesn't exist on type 'PullRequestChangedFile'
```

Github Authentification Output:
```
[Info  - 09:22:45.590] Getting user info...
[Info  - 09:22:46.204] Got account info!
```

Enterprise version is: v2.22.4

@kabel do you know why the above errors are happening? Is GHEP now out of sync with the extension again?

**Comment by bowendeng** [comment]
@jasonwilliams I had similar errors on our GHE instance. Based on [the change log](https://docs.github.com/en/graphql/overview/changelog#schema-changes-for-2020-09-23), these fields were introduced on 2020-09-23. I guess you have to upgrade the GHE to a version after that.

**Comment by kabel** [comment]
Opened a tracking issue at https://github.com/microsoft/vscode-pull-request-github/issues/2772 for the schema related errors. Gotta love GraphQL API version fragmentation. We'll see how hard it will be to keep up with Enterprise support pattern. My enterprise upgraded to GHE 3 and it's seems to be working fine.

**Comment by alexr00** [comment]
If you arrived here looking for GitHub Enterprise support in the GitHub Pull Requests and Issues extension, please see [the wiki](https://github.com/microsoft/vscode-pull-request-github/wiki). The TL;DR is that GitHub Enterprise is now tested and supported by the project maintainers.

**Pull request merged**
